### PR TITLE
allow plugins unload in supernova also

### PR DESCRIPTION
**Pull request by sonoro1234**

### CODE REVIEW

**Comment by sonoro1234** [comment]
usefull to unload DiskIO_UGens and avoid zombies.
Can be used in other future plugins unload

**Comment by timblechmann** [comment]
hmm, this should actually go to `close_handles` which is called from the destructor. however the handles are not closed (`dlclose` or `FreeLibrary`)  on purpose, most likely because some plugins tend not to clean up after themselves (spawn threads, but don't join them or the like). maybe that is, what you are trying to fix?

--

fwiw, i don't think that adding an "unload" handle is the correct way to solve this: instead of calling functions from an "unload" handle, one should create a global static object in the plugin, which will cleanly be destructed when the plugin is unloaded properly (dlcclose/freelibrary). in that case, there is no need to specify an "unload" symbol, but one can rely on the operating system to do the dirty cleanup.

**Comment by sonoro1234** [comment]
I tried to put that in close_handles but it is not called. At least from mingw-w64 compiler destructor is not automatically called.

**Comment by timblechmann** [comment]
that looks like a bug to me ... it should be destroyed in `~nova_server`, but it isn't ...

**Comment by sonoro1234** [comment]
~nova_server is called but not ~sc_ugen_factory because it is created with new 
sc_factory = new sc_ugen_factory;

**Comment by timblechmann** [comment]
that's the bug. it should be destroyed in `~nova_server`

**Comment by sonoro1234** [comment]
I have tried adding delete s_factory which calls FreeLibrary in win32 and dlclose otherwise. It is not enought to avoid zombie behaviour in DiskIO_UGens. It seems to need calling unload which calls gDiskIO.~DiskIOThread();

**Comment by timblechmann** [comment]
this looks like a bug in the mingw runtime. fwiw, it should be possible to use DllMain with DLL_PROCESS_DETACH, though.

compare: https://stackoverflow.com/questions/4496233/which-is-called-first-dllmain-or-global-static-object-constructor

**Comment by sonoro1234** [comment]
I will investigate mingw DllMain issue

**Comment by sonoro1234** [comment]
Just calling FreeLibrary is enought for gDiskIO.~DiskIOThread(); being called (so unload seems innecesary) but mThread.join() does not succed.

**Comment by timblechmann** [comment]
hmm, google tells me:
https://stackoverflow.com/questions/10441048/exit-thread-upon-deleting-static-object-during-unload-dll-causes-deadlock

so probably an 'unload' handle is the way to go ... sry for leading you on the wrong trail :/

**Comment by sonoro1234** [comment]
I have tried with gDiskIO as a pointer initialized in load() with new and there is no deadlock even if I dont use unload() or FreeLIbrary!!!
How can this be explained?

**Comment by timblechmann** [comment]
the thread is sleeping and never waked up. if it would wake up, it is going to crash. also it is going to crash if the thread is doing some work while the plugin is unloaded.

**Comment by sonoro1234** [comment]
waked up by gDiskIO->launchThread(); in load??

**Comment by timblechmann** [comment]
if the thread sleeps in  mDiskFifoHasData.WaitEach(), it won't crash. if the thread is anywhere else, it will

**Comment by sonoro1234** [comment]
But this is the same that happened before my commits. What is the relevant change?

**Comment by timblechmann** [comment]
the destructor is not called

**Comment by sonoro1234** [comment]
is called in
```
C_LINKAGE SC_API_EXPORT void unload(InterfaceTable *inTable)
{
	delete gDiskIO;
}
```

**Comment by timblechmann** [comment]
```
I have tried with gDiskIO as a pointer initialized in load() with new and there is no deadlock even if I dont use unload() or FreeLIbrary!!!
How can this be explained?
```

if you don't call the destructor on a pointer, the thread won't be joined. if the destructor is called by the run-time it apparently deadlocks or from DLLMain, it apparently deadlocks.

**Comment by sonoro1234** [comment]
That was just an experiment. In the actual version I call load at the beggining and unload and FreeLibrary at the end.
This should be ok?

File: server/supernova/sc/sc_ugen_factory.cpp [hunk 1]
```diff
@@ -279,10 +279,19 @@ void sc_ugen_factory::load_plugin ( boost::filesystem::path const & path )
     /* don't close the handle */
     return;
 }
-
+void sc_ugen_factory::unload_plugins(void)
```
Comment (timblechmann) [inline server/supernova/sc/sc_ugen_factory.cpp]: please call this in the beginning of close_handles (or merge both functions)

File: server/supernova/sc/sc_ugen_factory.hpp [hunk 2]
```diff
@@ -230,7 +230,7 @@ class sc_ugen_factory:
 
     void load_plugin_folder(boost::filesystem::path const & path);
     void load_plugin(boost::filesystem::path const & path);
-
+    void unload_plugins(void);
```
Comment (timblechmann) [inline server/supernova/sc/sc_ugen_factory.hpp]: private (or remove)

File: server/supernova/server/main.cpp [hunk 3]
```diff
@@ -383,5 +383,6 @@ int main(int argc, char * argv[])
     } else
         server.run_nonrt_synthesis(args);
 
+    sc_factory->unload_plugins();
```
Comment (timblechmann) [inline server/supernova/server/main.cpp]: unneeded when called from the destructor.

**Comment by timblechmann** [comment]
functionality is fine. just some minor cleanups to make everything self-contained.

thanks a lot!!

**Comment by sonoro1234** [comment]
Did minor cleanups.
ok now?

**Comment by timblechmann** [comment]
wonderful ... looks great now ... thanks a lot!

**Comment by bagong** [comment]
Builds on Windows 32-bit MinGW and 64-bit VS...

**Comment by sonoro1234** [comment]
playing now with internal server I see that World_New has
```
static bool gLibInitted = false;
		if (!gLibInitted) {
			InterfaceTable_Init();
			initialize_library(inOptions->mUGensPluginPath);
			initializeScheduler();
			gLibInitted = true;
		}
```
the purpose is that it is only executed the first time World_New is used.
Is there a good reason for that?

**Comment by sonoro1234** [comment]
The only reason I see is that initializeScheduler in SC_CoreAudio.cpp creates a thread that is never destroyed. So this is the code that should be run only once!!

**Comment by sonoro1234** [comment]
Instead of 
```
void initializeScheduler()
{
	syncOSCOffsetWithTimeOfDay();

	thread resyncThread(resyncThreadFunc);
	resyncThread.detach();
}
```
should be

```
thread *resyncThread = 0;
void initializeScheduler()
{
	syncOSCOffsetWithTimeOfDay();
        if(resyncThread == NULL){
	    resyncThread = new thread (resyncThreadFunc);
	    resyncThread->detach();
        }
}
```
Any thougths?

**Comment by sonoro1234** [comment]
Hi Tim and Rainer
I would need somebody trying the internal server. 
I can only try internal server with my Lua2SC client: It worked with boot and quit several times but had to apply last commit for avoiding crash.
(modifing World_New to allow initiailization each time would be cleaner for me)

**Comment by bagong** [comment]
Hi Victor and Tim,

I built with that commit (VS64 & MinGW32). The build went through fine, but unfortunately I don't get sound from the internal server yet. Not sure what's going on, I do get sound from the localhost server.

I am happy to test builds for you, but at some moment I might not be available. So I did the following: I gave you full access to my sc-repo on https://github.com/bagong/supercollider.git. The default branch is "winport" there. It is a cleaned up version of Lucas' Cornelisse Windows pull-request with a few, mainly cmake-related additions. If you push Windows-related changes to that branch things will be kept together, and at some point we could make a joint pull-request to SC master. The more important thing is this: any push on that branch will trigger a build in my build-system. This should generate a VS64 build and a MinGW32 build, create Installers (with git-hash in name) and move them to a shared Google-drive folder:

https://drive.google.com/folderview?id=0B7igZrWv7UxdUmxHdThYLUE2QXc&usp=sharing

The MinGW build is very big (330 MB) so it'll take a while to upload. I guess it's so big because it's a debug build. Release builds with MinGW currently cause a broken SC. Maybe because of optimizations that don't fit gcc4.9.2
The VS64 build is normal size.
I don't build SuperNova atm, but it shouldn't be a problem with the MinGW build, so if you're interested in that, just let me know.
Best
.r.

**Comment by sonoro1234** [comment]
Hi Rainer,
Are you using mingw from www.mingw.org/wiki/mingw (clasic)
or mingw-w64 from http://sourceforge.net/projects/mingw-w64/ with web page http://mingw-w64.yaxm.org/doku.php ?
I am using this second one (version mingw-builds 4.8.2 posix-dwarf) because it allows using C++11 for supernova

**Comment by bagong** [comment]
I am using MinGW as provided by msys2, this is the output of gcc -v:
```
Using built-in specs.
COLLECT_GCC=C:\msys\mingw32\bin\gcc.exe
COLLECT_LTO_WRAPPER=C:/msys/mingw32/bin/../lib/gcc/i686-w64-mingw32/4.9.2/lto-wrapper.exe
Target: i686-w64-mingw32
Configured with: ../gcc-4.9.2/configure --prefix=/mingw32 --with-local-prefix=/mingw32/local --build=i686-w64-mingw32 --host=i686-w64-mingw32 --target=i686-w64-mingw32 --with-native-system-header-dir=/mingw32/i686-w64-mingw32/include --libexecdir=/mingw32/lib --with-gxx-include-dir=/mingw32/include/c++/4.9.2 --enable-bootstrap --with-arch=i686 --with-tune=generic --enable-languages=c,lto,c++,objc,obj-c++,fortran,ada --enable-shared --enable-static --enable-libatomic --enable-threads=posix --enable-graphite --enable-fully-dynamic-string --enable-libstdcxx-time=yes --disable-libstdcxx-pch --disable-libstdcxx-debug --enable-cloog-backend=isl --enable-version-specific-runtime-libs --disable-cloog-version-check --disable-isl-version-check --enable-lto --enable-libgomp --disable-multilib --enable-checking=release --disable-rpath --disable-win32-registry --disable-nls --disable-werror --disable-symvers --with-libiconv --with-system-zlib --with-gmp=/mingw32 --with-mpfr=/mingw32 --with-mpc=/mingw32 --with-isl=/mingw32 --with-cloog=/mingw32 --with-pkgversion='Rev5, Built by MSYS2 project' --with-bugurl=http://sourceforge.net/projects/msys2 --with-gnu-as --with-gnu-ld --disable-sjlj-exceptions --with-dwarf2
Thread model: posix
gcc version 4.9.2 (Rev5, Built by MSYS2 project)
```
My considerations as to going for this were:
- msys2 provides all packages we need, built in a single environment
- they provide pacman, so you install packages like in Arch Linux
- it provides everything in 32- and 64-bit versions
- they are very quick with gcc & Qt-related packages and are the only ones I know that provide 64-bit Qt-mingw
- it's a dynamic project setting out to fill the gap left by too slowly developing msys & cygwin

It's a continuation of mingwbuilds and a fork of cygwin. On the downside: it's not very mature yet. But creating a comprehensive build environment for SC with them is done in a breeze compared to the past. And no more telling cmake where dependencies are. Thread model seems to be the same as the one you got (I sometimes see remarks in the SC sources that point towards expecting windows thread model?). C++11 no prob with 4.9.2 and I have built SuperNova in this environment, I just didn't add it to the automatic build as I thought it's not needed. But if you're interested I am happy to add it.
There seems to be a problem with compiler optimizations with 4.9.2. In Release and RelWithDebInfo SC breaks at startup. That was not the case with gcc 4.8.2

As to internal server: maybe this observation helps: when I execute ().play - just no sound and the normal event output in the post window. When I execute {SinOsc.ar(mul:0.1)}.play, I get a warning: `WARNING: server 'localhost' not running`. So the system doesn't know that the internal server has been booted.

**Comment by bagong** [comment]
Hey Victor,

sorry, sorry, sorry. The internal server does work! I forgot to set Server.default to Server.internal!

embarrassed
.r.

**Comment by sonoro1234** [comment]
great!!
With the static bool gLibInitted = false; issue I expect it to initialize and deinitialize (World_New and World_Clear) correctly only the first time but to fail the second time you boot it (for example in DiskIO_UGens)
Trying your build I got a weird msvc runtime error opening scide so I could not test

**Comment by bagong** [comment]
I am just looking into the runtime/dll-stuff, had to find another computer first...

**Comment by sonoro1234** [comment]
Hi Rainer with your build I tested sclang.exe:
```
init_OSC
empty
compiling class library (debug build)...
initPassOne started
This application failed to start because it could not find or load the Qt platfo
rm plugin "windows".

Reinstalling the application may fix this problem.

This application has requested the Runtime to terminate it in an unusual way.
Please contact the application's support team for more information.


```

**Comment by timblechmann** [comment]
you need to deploy qt's windows platform plugins

**Comment by bagong** [comment]
Yea, I am on my way

**Comment by bagong** [comment]
It's done. The new versions (including Victor's latest commit) build nice with Mingw32 and VS studio. windows platform plugins are included. Would be great if somebody had a second look. SuperNova is included in the MinGW build. It doesn't compile on current VS.
Best
.r.

**Comment by sonoro1234** [comment]
Have tried bootting and quitting several times but find difficult to try other things because each time I do a selection sclang.exe crashes

**Comment by sonoro1234** [comment]
Have tried with several evaluate files instead of evaluate selection. DiskIO_UGens works after several boot quit cicles.
Somebody with coreaudio (osx?) should try the void initializeScheduler() issue (I expect several resync threads after several boot quit cycles) 
Or perhaps is better to do proposed solution first?

**Comment by sonoro1234** [comment]
Hi Tim,
I cant compile neither test that (I am on win32)
Is last commit ok?

**Comment by bagong** [comment]
Victor, yea, that's the known show-stopper bug on Windows. Don't drag with a mouse and use the keyboard as little as possible. Paste your code, select everything and hit ctrl-enter. That should work, or file evaluate.
I can test your stuff on OS X if nobody else jumps in, but give me some test code :)

**Comment by sonoro1234** [comment]
apply last commit and test compilation.
testing function is more difficult:
As possible issues are with clock resync you should get late messages if sampleclock is faster or delayed response if sampleclok is slower than systemclock

**Pull request closed without merging**